src/org/rascalmpl/library/util/ErrorRecovery.rsc

```diff
@@ -51,3 +51,33 @@ This filter removes error trees until no ambiguities caused by error recovery ar
 Note that regular ambiguous trees remain in the parse forest unless `allowAmbiguity` is set to false in which case an error is thrown.
 }
 java Tree disambiguateErrors(Tree t, bool allowAmbiguity=true);
+
+@javaClass{org.rascalmpl.library.util.ErrorRecovery}
+@synopsis{Check if two parse trees are equal. For "normal" parse trees `==` will suffice, but for highly ambiguous (error) trees we need an
+equality check that takes advantage of the sharing present in those trees.
+Note that this equality check only takes into account the offset and length fields of source locations. All other fields are ignored.}
+java bool treeEquality(Tree t1, Tree t2);
+
+int nodeCount(appl(_, args)) {
+    int count = 1;
+    for (Tree arg <- args) {
+        count += nodeCount(arg);
+    }
+    return count;
+}
+
+int nodeCount(amb(alts)) {
+    int count = 1;
+    for (Tree alt <- alts) {
+        count += nodeCount(alt);
+    }
+    return count;
+}
+
+int nodeCount(char(_)) = 1;
+
+int nodeCount(cycle(_,_)) = 1;
+
+@javaClass{org.rascalmpl.library.util.ErrorRecovery}
```

Comment:
The Rascal declaration of the `parseTree2Dot` function will be removed in the next PR. The reason is that the resulting graph is not very clean and specific to the debugging of memoization so it is not all that useful in general in its current form.
We will keep the Java implementation around because we might want to improve it in the future or we mightg need more memoization debugging.